sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -50,7 +50,14 @@ trait InvokeLike extends Expression with NonSQLExpression {
 
   def propagateNull: Boolean
 
-  protected lazy val needNullCheck: Boolean = propagateNull && arguments.exists(_.nullable)
+  def propagateNullForPrimitive: Boolean
```

Comment:
This reminds me of the rule `HandleNullInputsForUDF`.

For primitive inputs, I don't think we have a choice and we must propagate null (`isPositive(int i)` can't handle null values). Can we detect it automatically instead of adding a new boolean flag?

Reply:
OK. Let me remove the flag then, and add an extra comment to `propagateNull`.

One issue with this approach is that it only applies to the magic method path, while for `produceResult` users will need to handle the null primitive values explicitly. I think we'll need to document this more carefully, otherwise it could cause confusion.

Reply:
yea, I agree. If users want to handle null by themselves, they should use boxed primitive types as the UDF parameter type.